sdk/lib/conutils/pager.c

```diff
@@ -31,7 +33,115 @@
 // Temporary HACK
 #define CON_STREAM_WRITE    ConStreamWrite
 
+#define CP_SHIFTJIS 932  // Japanese Shift-JIS
+#define CP_HANGUL   949  // Korean Hangul/Wansung
+#define CP_JOHAB    1361 // Korean Johab
+#define CP_GB2312   936  // Chinese Simplified (GB2312)
+#define CP_BIG5     950  // Chinese Traditional (Big5)
 
+/* IsFarEastCP(CodePage) */
+#define IsCJKCodePage(CodePage) \
+    ((CodePage) == CP_SHIFTJIS || (CodePage) == CP_HANGUL || \
+  /* (CodePage) == CP_JOHAB || */ \
+     (CodePage) == CP_BIG5     || (CodePage) == CP_GB2312)
+
+static inline INT GetWidthOfCharCJK(UINT nCodePage, WCHAR ch)
+{
+    INT ret = WideCharToMultiByte(nCodePage, 0, &ch, 1, NULL, 0, NULL, NULL);
+    if (ret == 0)
+        ret = 1;
+    else if (ret > 2)
+        ret = 2;
+    return ret;
+}
+
+static BOOL CALLBACK
+ConDefaultPagerLine(PCON_PAGER Pager, LPCWSTR line, DWORD cch, DWORD *pdwFlags)
+{
+    CON_STREAM_WRITE(Pager->Screen->Stream, line, cch);
+    return TRUE;
+}
+
+static VOID CALLBACK
+ConCallPagerLine(PCON_PAGER Pager, LPCWSTR line, DWORD cch, DWORD *pdwFlags)
+{
+    if (!Pager->PagerLine || !(*Pager->PagerLine)(Pager, line, cch, pdwFlags))
+        (*Pager->DefPagerLine)(Pager, line, cch, pdwFlags);
+}
+
+static BOOL CALLBACK ConPagerDefaultAction(PCON_PAGER Pager)
+{
+    PCTCH TextBuff = Pager->TextBuff;
+    DWORD ich = Pager->ich;
+    DWORD cch = Pager->cch;
+    DWORD iColumn = Pager->iColumn;
+    DWORD iLine = Pager->iLine;
+    DWORD lineno = Pager->lineno;
+    DWORD ScreenColumns = Pager->ScreenColumns;
+    DWORD ScrollRows = Pager->ScrollRows;
+    DWORD ichLast = ich;
+    UINT nWidthOfChar = 1;
+    UINT nCodePage;
+    BOOL IsCJK = FALSE;
+    BOOL IsDoubleWidthCharTrailing = FALSE;
+    DWORD dwFlags = 0;
+
+    if (ich >= cch)
+        return TRUE;
+
+    nCodePage = GetConsoleOutputCP();
+    IsCJK = IsCJKCodePage(nCodePage);
+
+    for (; ich < cch && iLine < ScrollRows; ++ich)
+    {
+        Pager->lineno = lineno;
+        if (IsCJK)
+        {
+            nWidthOfChar = GetWidthOfCharCJK(nCodePage, TextBuff[ich]);
+            IsDoubleWidthCharTrailing = (nWidthOfChar == 2) &&
+                                        (iColumn + 1 == ScreenColumns);
```

Comment:
```suggestion
                                        (iColumn + 1 >= ScreenColumns);
```

Reply:
`iColumn + 2 == ScreenColumns` is not the case of this.

Reply:
With the `>=` this makes the code more robust.

Reply:
If `iColumn + 2 == ScreenColumns`, then the double-width character will fit the screen edge and it doesn't overlap.

Reply:
ok thanks for the explanation.